modules/javafx.graphics/src/main/java/com/sun/javafx/application/preferences/PlatformPreferences.java

```diff
@@ -0,0 +1,229 @@
+import com.sun.javafx.collections.MapListenerHelper;
+import javafx.application.Appearance;
+import javafx.application.Platform;
+import javafx.beans.InvalidationListener;
+import javafx.beans.property.ReadOnlyObjectProperty;
+import javafx.collections.MapChangeListener;
+import javafx.scene.paint.Color;
+import javafx.scene.paint.Paint;
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Contains the implementation of a read-only map of platform preferences.
+ * <p>
+ * When the operating system signals that a preference has changed, the mappings are updated
+ * by calling the {@link #update(Map)} method.
+ */
+public class PlatformPreferences extends AbstractMap<String, Object> implements Platform.Preferences {
+
+    /**
+     * Contains mappings from platform-specific keys to well-known keys, which are used
+     * in the implementation of the property-based API in {@link PreferenceProperties}.
+     */
+    final Map<String, String> wellKnownKeys;
+    final Map<String, Object> effectivePreferences = new HashMap<>();
+    final Map<String, Object> unmodifiableEffectivePreferences = Collections.unmodifiableMap(effectivePreferences);
+    final PreferenceProperties properties = new PreferenceProperties(this);
+
+    private MapListenerHelper<String, Object> helper;
+
+    public PlatformPreferences(Map<String, String> wellKnownKeys) {
+        this.wellKnownKeys = wellKnownKeys;
+    }
+
+    @Override
+    public Set<Entry<String, Object>> entrySet() {
+        return unmodifiableEffectivePreferences.entrySet();
+    }
+
+    @Override
+    public Object remove(Object key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void addListener(InvalidationListener listener) {
+        helper = MapListenerHelper.addListener(helper, listener);
+    }
+
+    @Override
+    public void removeListener(InvalidationListener listener) {
+        helper = MapListenerHelper.removeListener(helper, listener);
+    }
+
+    @Override
+    public void addListener(MapChangeListener<? super String, ? super Object> listener) {
+        helper = MapListenerHelper.addListener(helper, listener);
+    }
+
+    @Override
+    public void removeListener(MapChangeListener<? super String, ? super Object> listener) {
+        helper = MapListenerHelper.removeListener(helper, listener);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> Optional<T> getValue(String key, Class<T> type) {
+        Objects.requireNonNull(key, "key cannot be null");
+        Object value = effectivePreferences.get(key);
+
+        if (value == null) {
+            return Optional.empty();
+        }
+
+        if (type.isInstance(value)) {
+            return Optional.of((T)value);
+        }
+
+        throw new IllegalArgumentException(
```

Comment:
should this behavior be documented?
there is no mention of an exception in case of type mismatch in the base class.

also, do we want to have some kind of trivial conversion implemented such as int -> long -> double ?  
or not?

Reply:
It is documented in the `Platform.Preferences` interface, which is implemented by this class:

```
        /**
         * Returns the value to which the specified key is mapped.
         *
         * @param <T> the type of the value
         * @param key the key
         * @param type the type of the value
         * @throws NullPointerException if {@code key} is null
         * @throws IllegalArgumentException if the key is not mapped to a {@code type} instance
         * @return the value to which the key is mapped, or {@code Optional.empty()}
         *         if no mapping exists for the specified key
         */
        <T> Optional<T> getValue(String key, Class<T> type);
```

I'm a bit hesitant regarding an automatic conversion. What would be the use case for this?

Reply:
one [possible] use case is when different versions of the native platform have different types (int32/int64 ?).
or, if the property value is String with the semantics of an integer.

Of course, the app dev can always request an Object and parse that.

Reply:
If we add this, we'd also need to decide whether we only want to support lossless conversions, or if lossy conversions are acceptable as well. Since this will add a considerable amount of complexity to this PR, I suggest we don't to it at this time. If there's a need, we can always add it later.

Reply:
after some thought, I would say conversion is not needed - should that unlikely scenario happen, the app devs could use getValue(.., Object.class);